src/sai/mod.rs

```diff
@@ -122,34 +144,101 @@ macro_rules! sai_hal {
+                        Event::LateFrameSync        => ch.im.write(|w| w.lfsdetie().set_bit()),
+                    }
+                }
 
-                /// Start listening for `event`
-                pub fn listen(&mut self, event: Event) {
+                /// Stop listening for `event` on a given `channel`
+                pub fn unlisten(&mut self, channel: SaiChannel, event: Event) {
+                    let ch = match channel {
+                        SaiChannel::ChannelA => &self.rb.cha,
+                        SaiChannel::ChannelB => &self.rb.chb,
+                    };
                     match event {
-                        Event::Data => {
-                            // Enable FIFO request interrupt
-                            self.rb.cha.im.modify(|_, w| w.freqie().set_bit());
-                        }
+                        Event::Overdue              => ch.im.write(|w| w.ovrudrie().clear_bit()),
+                        Event::Muted                => ch.im.write(|w| w.mutedetie().clear_bit()),
+                        Event::WrongClock           => ch.im.write(|w| w.wckcfgie().clear_bit()),
+                        Event::Data                 => ch.im.write(|w| w.freqie().clear_bit()),
+                        Event::AnticipatedFrameSync => ch.im.write(|w| w.afsdetie().clear_bit()),
+                        Event::LateFrameSync        => ch.im.write(|w| w.lfsdetie().clear_bit()),
                     }
                 }
 
-                /// Stop listening for `event`
-                pub fn unlisten(&mut self, event: Event) {
+                /// Clears interrupt flag `event` on the `channel`
+                ///
+                /// Note: Event::Data is accepted but does nothing as that flag is cleared by reading/writing data
+                pub fn clear_irq(&mut self, channel: SaiChannel, event: Event) {
+                    let ch = match channel {
+                        SaiChannel::ChannelA => &self.rb.cha,
+                        SaiChannel::ChannelB => &self.rb.chb,
+                    };
                     match event {
-                        Event::Data => {
-                            // Enable FIFO request interrupt
-                            self.rb.cha.im.modify(|_, w| w.freqie().clear_bit());
-                        }
+                        Event::Overdue              => ch.clrfr.write(|w| w.covrudr().set_bit()),
+                        Event::Muted                => ch.clrfr.write(|w| w.cmutedet().set_bit()),
+                        Event::WrongClock           => ch.clrfr.write(|w| w.cwckcfg().set_bit()),
+                        Event::Data                 => (), // Cleared by reading/writing data
+                        Event::AnticipatedFrameSync => ch.clrfr.write(|w| w.cafsdet().set_bit()),
+                        Event::LateFrameSync        => ch.clrfr.write(|w| w.clfsdet().set_bit()),
```

Comment:
These are correctly `write` though

Reply:
Got it. Did a pass on the modules I touched for this and fixed them.